kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java

```diff
@@ -0,0 +1,455 @@
+          consumerWrapper.commitOffsets(item.getRight());
+        }
+        iter.remove();
+      }
+    }
+  }
+
+  @Override
+  public void emitTuples()
+  {
+    int count = consumerWrapper.messageSize();
+    if (maxTuplesPerWindow > 0) {
+      count = Math.min(count, maxTuplesPerWindow - emitCount);
+    }
+    for (int i = 0; i < count; i++) {
+      Pair<String, ConsumerRecord<K, V>> tuple = consumerWrapper.pollMessage();
+      ConsumerRecord<K, V> msg = tuple.getRight();
+      emitTuple(tuple.getLeft(), msg);
+      AbstractKafkaPartitioner.PartitionMeta pm = new AbstractKafkaPartitioner.PartitionMeta(tuple.getLeft(),
+          msg.topic(), msg.partition());
+      offsetTrack.put(pm, msg.offset());
+    }
+    emitCount += count;
+  }
+
+  protected abstract void emitTuple(String cluster, ConsumerRecord<K, V> msg);
+
+  @Override
+  public void beginWindow(long wid)
+  {
+    emitCount = 0;
+    currentWindowId = wid;
+  }
+
+  @Override
+  public void endWindow()
+  {
+    // copy current offset track to history memory
+    Map<AbstractKafkaPartitioner.PartitionMeta, Long> offsetsWithWindow = new HashMap<>(offsetTrack);
+    offsetHistory.add(Pair.of(currentWindowId, offsetsWithWindow));
+
+    //update metrics
+    metrics.updateMetrics(clusters, consumerWrapper.getAllConsumerMetrics());
+  }
+
+
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    applicationName = context.getValue(Context.DAGContext.APPLICATION_NAME);
+    consumerWrapper.create(this);
+    metrics = new KafkaMetrics(metricsRefreshInterval);
+  }
+
+  @Override
+  public void teardown()
+  {
+
+  }
+
+  private void initPartitioner()
+  {
+    if (partitioner == null) {
+      logger.info("Initialize Partitioner");
+      switch (strategy) {
+        case ONE_TO_ONE:
+          partitioner = new OneToOnePartitioner(clusters, topics, this);
+          break;
+        case ONE_TO_MANY:
+          partitioner = new OneToManyPartitioner(clusters, topics, this);
+          break;
+        case ONE_TO_MANY_HEURISTIC:
+          throw new UnsupportedOperationException("Not implemented yet");
+        default:
+          throw new RuntimeException("Invalid strategy");
+      }
+      logger.info("Actual Partitioner is {}", partitioner.getClass());
+    }
+
+  }
+
+  @Override
+  public Response processStats(BatchedOperatorStats batchedOperatorStats)
+  {
+    long t = System.currentTimeMillis();
+    if (repartitionInterval < 0 || repartitionCheckInterval < 0 ||
+        t - lastCheckTime < repartitionCheckInterval || t - lastRepartitionTime < repartitionInterval) {
+      // return false if it's within repartitionCheckInterval since last time it check the stats
+      Response response = new Response();
+      response.repartitionRequired = false;
+      return response;
+    }
+
+    try {
+      logger.debug("Process stats");
+      initPartitioner();
+      return partitioner.processStats(batchedOperatorStats);
```

Comment:
3 possibilities here
1. within interval, the operator shouldn't not try repartition
2. exceed interval, the partitioner will always try to repartition based on latest kafka partition info. If there is no kafka partition change, it will return same collection of partition instances and AM will not try to repartition the operator.
3. exceed interval and there is kafka partition change(ex. new kafka partition added at runtime) it will try to repartition the operator.